### PR TITLE
Merging this sprint's transactions work

### PR DESCRIPTION
**Pull request by shellajt**
- Setup database architecture
- implemented backend class structure
- Setup SQLiteDatabase to implement IDatabase
- Added ability to add/edit/delete transactions

### CODE REVIEW

**Comment by taylerhow** [comment]
Looks good to me! But I can't really review it, since I wrote a lot of it.

File: src/main/java/ledger/database/storage/SQLiteDatabase.java [hunk 1]
```diff
@@ -0,0 +1,988 @@
...
+        try {
+            database.close();
+        } catch (SQLException e) {
+            throw new StorageException("Exception while shutting down database.", e);
+        }
+    }
+
+    // Basic CRUD functionality
+    @Override
+    public void insertTransaction(Transaction transaction) throws StorageException {
```
Comment (LadyMopar) [inline src/main/java/ledger/database/storage/SQLiteDatabase.java]: We could probably refactor this method and break out some of the pieces to make it shorter. This isn't something we need to worry about right now, however.

File: src/main/java/ledger/database/storage/SQLiteDatabase.java [hunk 2]
```diff
@@ -0,0 +1,988 @@
...
+
+            deleteNoteForTransactionID(transaction.getId());
+            deleteAllTagToTransForTransactionID(transaction.getId());
+
+        } catch (java.sql.SQLException e) {
+            throw new StorageException("Error while deleting transaction", e);
+        }
+    }
+
+    @Override
```
Comment (LadyMopar) [inline src/main/java/ledger/database/storage/SQLiteDatabase.java]: Same here.

File: src/main/java/ledger/database/storage/SQLiteDatabase.java [hunk 3]
```diff
@@ -0,0 +1,988 @@
...
+            throw new StorageException("Error while getting Account by Name", e);
+        }
+    }
+
+    private Tag getTagForNameAndDescription(String tagName, String tagDescription){
+        try {
+            PreparedStatement stmt = database.prepareStatement("SELECT * FROM TAG WHERE TAG_NAME=? AND TAG_DESC=?");
+            stmt.setString(1, tagName);
+            stmt.setString(2, tagDescription);
+            ResultSet rs =stmt.executeQuery();
```
Comment (LadyMopar) [inline src/main/java/ledger/database/storage/SQLiteDatabase.java]: Need a space between = and stmt

File: src/main/java/ledger/database/storage/SQLiteDatabase.java [hunk 4]
```diff
@@ -0,0 +1,988 @@
...
+            }
+
+            Payee existingPayee = getPayeeForNameAndDescription(transaction.getPayee().getName(), transaction.getPayee().getDescription());
+            if (existingPayee != null) {
+                stmt.setInt(6, existingPayee.getId());
+            } else {
+                insertPayee(transaction.getPayee());
+                Payee insertedPayee = getPayeeForNameAndDescription(transaction.getPayee().getName(), transaction.getPayee().getDescription());
+                stmt.setInt(6, insertedPayee.getId());
+
```
Comment (LadyMopar) [inline src/main/java/ledger/database/storage/SQLiteDatabase.java]: there's an extra space here

File: src/main/java/ledger/database/storage/SQLiteDatabase.java [hunk 5]
```diff
@@ -0,0 +1,988 @@
...
+            deleteAllTagToTransForTransactionID(transaction.getId());
+            for (Tag currentTag : transaction.getTagList()) {
+                Tag existingTag = getTagForNameAndDescription(currentTag.getName(), currentTag.getDescription());
+                if (existingTag != null) {
+                    insertTagToTrans(existingTag.getId(), transaction.getId());
+                } else {
+                    insertTag(currentTag);
+                    Tag insertedTag = getTagForNameAndDescription(currentTag.getName(), currentTag.getDescription());
+                    insertTagToTrans(insertedTag.getId(), transaction.getId());
+
```
Comment (LadyMopar) [inline src/main/java/ledger/database/storage/SQLiteDatabase.java]: another extra line

File: src/main/java/ledger/database/storage/SQLiteDatabase.java [hunk 6]
```diff
@@ -0,0 +1,988 @@
...
+            ")";
+    private static final String tableTagToPayee = "CREATE TABLE IF NOT EXISTS TAG_TO_PAYEE " +
+            "(TTPE_TAG_ID INT           NOT NULL, " +
+            "TTPE_PAYEE_ID INT          NOT NULL, " +
+            "FOREIGN KEY(TTPE_TAG_ID) REFERENCES TAG(TAG_ID), " +
+            "FOREIGN KEY(TTPE_PAYEE_ID) REFERENCES PAYEE(PAYEE_ID)" +
+            ")";
+
+    private Connection database;
+
```
Comment (LadyMopar) [inline src/main/java/ledger/database/storage/SQLiteDatabase.java]: extra line

**Comment by shellajt** [comment]
Suggested edits will be implemented during the next sprint

**Pull request merged**